### PR TITLE
QL: Retrieve case sensitivity from configuration

### LINKED ISSUE #59899: QL: Push isCaseSensitive handling down to Configuration
Related to #54411

There was a repeated pattern we kept running into when adding case handling (sensitivity/insensitivity toggles) to QL and EQL functions. We generally would make the relevant QL functions case sensitive, then we would subclass and change the handling.

It's an odd pattern, and we could instead make the Configuration provide `isCaseSensitive()`, defaulting to `true`, and refer to it when determine which mode to handle a function. I think it simplifies the code a bit and makes it easier to extend to new functions.

### PR DESCRIPTION
**Pull request by rw-access**
Closes #59899

Instead of subclassing StartsWith and following that pattern for other functions, it makes more sense to push the `isCaseSensitive()` method down to QL's `Configuration`, which we can use more generically.

I think this will make it easier to add case sensitive/insensitive handling in QL. It'll also simplify the work for #56752 and #56749.

### CODE REVIEW

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/session/Configuration.java [hunk 1]
```diff
@@ -38,4 +40,8 @@ public String clusterName() {
     public String username() {
         return username;
     }
+
```
Comment (matriv) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/session/Configuration.java:43]: Needs `@Override`

**Review by matriv — COMMENTED** [review]
The idea looks good to me.

- Imho, the `CaseAware` iface is superfluous.
- I would make the `isCaseSensitive()` abstract on ql's Configuration so that is more clear that SqlConfiguration returns always true, whereas eql Configuration uses the value passed as param by the user.

**Comment by matriv** [comment]
@rw-access My suggestion was to make the method `isCaseSensitive()` abstract on ql's Configuration and not the class itself.

**Review by astefan — COMMENTED** [review]
There is one aspect that doesn't fit in this picture and I tend to disagree with this approach: why would the case sensitivity be a feature of QL in general, when only EQL is the one that really needs it?

Also, `StartsWith` is an exception. It lives in QL because both projects (sql and eql) make use of it. The same approach shouldn't be used for other eql functions.

**Comment by rw-access** [comment]
> There is one aspect that doesn't fit in this picture and I tend to disagree with this approach: why would the case sensitivity be a feature of QL in general, when only EQL is the one that really needs it?
>
> Also, StartsWith is an exception. It lives in QL because both projects (sql and eql) make use of it. The same approach shouldn't be used for other eql functions.

@astefan every QL function with string comparisons needs to support case insensitivity for QL, so StartsWith isn't the exception. Right now, it's an an exception because it's the only one we've completed so far. The other functions are: Like (Wildcard), RLike (Match), Equals, NotEquals, LessThan, LessThanOrEqual, GreaterThanOrEqual and In.

To me, pushing the case sensitivity toggle into QL makes a lot of sense. The alternative is that we write new functions, pipes and processors in EQL, which isn't worth the effort in my opinion. Another pro, by pushing it into QL, is that case-sensitive languages like SQL don't have to use the functionality. But if we have another query language that's case-insensitive, then we already have all the pieces we need, and just need to update `isCaseSensitive` on its `Configuration`. Maybe EQL will remain the only function that use case-insensitivity, but I still think it's a lot less effort to add the toggles to QL than rewrite or subclass functions, pipes and processors.

**Comment by rw-access** [comment]
@elasticmachine update branch

**Comment by costin** [comment]
I see more negatives than positives by pushing this concept into QL.

Conceptually, case insensitivity is a concept that is EQL specific, SQL clearly doesn't have it. It's also a loose concept since it applies not only to explicitly defined string functions (StartWith, etc..) but as we've discovered, to operators which now have to become aware of not just their type but also this extra parameter.
I'd rather restrict this behavior as much as possible due to its greedy nature.

Implementation wise, I see disadvantages. We already have code that is case sensitive and need to add code/tests for case insensitivity. Pushing that down to SQL is a liability since SQL functions will be case sensitive - we assume that code path will not be executed but then again... bugs prove us expectation != reality.

Let's talk about this more in our next meeting.

**Comment by costin** [comment]
Not needed anymore - closed by https://github.com/elastic/elasticsearch/issues/62949.

**Pull request closed without merging**